### ISSUE TITLE
Streamlining the installation and usage process

### ISSUE THREAD

**Issue by samuela** [princeton-nlp/SWE-agent#66]
(posting at @ofirpress's request)

The current installation process requires a number of steps, and requires a non-trivial investment for non-contributors. For example, someone from discord recently mentioned that this process took them 2 days. 
 
Ideally swe-agent would have a single CLI command that anyone could run from any machine, regardless of whether they have cloned the repo or downloaded any dependencies:
```
$ swe-agent \
  --github-issue=https://github.com/princeton-nlp/SWE-agent/issues/57 \
  --model=gpt4 \
  --openai_api_key=... \
  --cost-limit=2.00
```
  
# Design possibilities
Note that none of these solutions are mutually exclusive.
 
## Docker container
**Usage:** `docker run princeton-nlp/swe-agent ...`
 
In this solution, we would publish a Docker container with `CMD` set to `python run.py`.
 
Considerations:
* Would require the user to have Docker installed before running the command.
* Would involve docker-in-docker, which is do-able but brings its own subtleties.
 
## Nix flake
**Usage:** `nix run github:princeton-nlp/SWE-agent`
 
In this solution, we would package swe-agent as a [nix](https://nixos.org/) [flake](https://zero-to-nix.com/concepts/flakes). This has the advantage of increasing reproducibility and obviating the need for the current `setup.py`/`pyproject.toml` situation.
 
Considerations:
* Would require the user to have Nix, and possibly Docker, installed before running the command. While nix has packaged the Docker CLI, the Docker daemon will likely require a native installation on non-NixOS operating systems.
* Pro: would not require docker-in-docker
* Pro: dependency pulling and caching would be handled automatically by Nix
 
## curl-to-shell
**Usage:** `curl https://github.com/princeton-nlp/SWE-agent/blob/main/run.sh | sh`
 
Here we would be following in the footsteps of [rustup](https://rustup.rs/), [the Nix installer](https://nixos.org/download/), and friends.
 
Considerations
* Pro: would involve minimal changes to the existing packaging of SWE-agent
* Pro: would require effectively no pre-requisites
* Con: would require prompting for sudo to install Docker on Linux, and may not be possible on macOS
* Con: would require implementing logic to query, download, and cache this repo and its dependencies on the user's local machine. This would effectively be a reimplementation of package management logic that already exists in Nix/Docker.
* Con: curl-to-sh is commonly chided for its [security implications](https://0x46.net/thoughts/2019/04/27/piping-curl-to-shell/). Some people are reluctant to run curl-to-sh commands.

**Comment by ofirpress** [comment]
Thank you Sam. @klieret what do you think?

**Comment by klieret** [comment]
Thanks for the conclusive writeup!

> The current installation process requires a number of steps, and requires a non-trivial investment for non-contributors. For example, someone from discord recently mentioned that this process took them 2 days.

Getting feedback on what steps took them so long would be important! (if you know what they were struggling with). Can you ask them to open issues here? Would be much better as a knowledge base.

## Docker

> Would involve docker-in-docker, which is do-able but brings its own subtleties.

I will look into this today, but I don't think this is necessarily true. There seems to be a distinction between spawning child containers and sibling containers and we'd probably want the latter (I was reading [this blog post about it](https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/)). 

## setup.py & co

> This has the advantage of increasing reproducibility and obviating the need for the current setup.py/pyproject.toml situation.

Could you elaborate? I think this repository currently has the old-school setup.py rather than going full `pyproject.toml`, but it sounds like you're talking about a more fundamental flaw? 

Or are you in general talking about python dependency hell and the like? This is indeed something we could address, but I think there's two different philosophies here: using this project as a library (leaving most of the dependencies unpinned to increase interoperability) vs using this project as an app (pinning everything down for reproducibility). We could do both to some extend though. 

## Nix

Is that something that's used frequently in this field? I've heard of nix as a new package manager for UNIX systems, but I honestly haven't encountered it much previously (and no distribution other than NixOS uses it as a default package manager). Do you have more information to back up its usefulness?

## curl-to-shell

This is equivalent to just have a shell script that installs everything right? So it would ultimately boil down to the same set of instructions we currently have except packaged in a single script. So if people took 2 days to install everything so far, I imagine the script would not run out of the box for the exact same reasons. 

tl;dr: not a solution

**Comment by samuela** [comment]
> I will look into this today, but I don't think this is necessarily true. There seems to be a distinction between spawning child containers and sibling containers and we'd probably want the latter (I was reading [this blog post about it](https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/)).

This made me realize: swe-agent could avoid the docker-in-docker issue entirely by just using the initial Docker container. Since it's in a container anyways, there's no need to containerize further.

> Could you elaborate? I think this repository currently has the old-school setup.py rather than going full pyproject.toml, but it sounds like you're talking about a more fundamental flaw?

Yes, that's also my understanding. I don't see `setup.py` or pyproject as a fundamental flaw, but rather I list this as an auxiliary benefit -- I'm not sure if the project is currently suffering from Python packaging hell, but the Nix route would avoid any such issues in the future.

> Is that something that's used frequently in this field? [..] Do you have more information to back up its usefulness?

Nix is the largest and most up-to-date package manager [[1](https://www.reddit.com/r/NixOS/comments/zp95a2/so_according_to_repology_nix_has_an_insane_lead/), [2](https://repology.org/repositories/statistics/total), [3](https://repology.org/repositories/graphs)]. It is the default package manager for NixOS, but it is also popular (perhaps even moreso!) as a standalone package manager -- along the lines of homebrew, apt-get, etc -- that works on macOS and all Linux OSs.

One of Nix's value propositions is reproducibility: every package is hashed according to exactly how it was built and the hashes of its dependencies. This feature is especially popular among folks interested in reproducible research. 

Disclaimer: I am a Nix user and maintainer, but I can confirm that Nix would help to streamline the zero-to-running process.

> This is equivalent to just have a shell script that installs everything right? So it would ultimately boil down to the same set of instructions we currently have except packaged in a single script.

Yes, but also one that runs swe-agent. This would be intended as UX improvement, and would not require changing the packaging.

> So if people took 2 days to install everything so far, I imagine the script would not run out of the box for the exact same reasons.

Why? If a script can handle the installation and execution process end-to-end, wouldn't that solve the issue?

**Comment by klieret** [comment]
Thanks for the additional information!

I'll look into using Docker today and will report back

**Comment by zachblume** [comment]
Docker-compose comes with docker, can spawn two sibling containers that share networking, allow service-to-service communication, would avoid docker-in-docker, and allows for a single command setup by defining a docker-compose.yml and a terminal command like `docker compose up -d`

**Comment by klieret** [comment]
OK, so running `run.py` in a docker container actually works relatively easily. I can take a look at `docker compose` if that further simplifies things.

**Comment by zachblume** [comment]
E.g., Supabase's local development flow worked this way at first:
[https://supabase.com/docs/guides/self-hosting/docker](https://supabase.com/docs/guides/self-hosting/docker)

**Comment by samuela** [comment]
For an example of a seemingly-reasonable person struggling to follow the current installation instructions, see https://www.youtube.com/watch?v=9-JBHGlYEBI.

> I can take a look at docker compose if that further simplifies things.

docker compose would require cloning the repo IIUC

**Comment by zachblume** [comment]
```
curl -s https://example.com/path/to/docker-compose.yml | docker-compose -f - up
```
But it may easier to offer config with a cloned repo, e.g. the Supabase example above. I think there's strong arguments to curl-to-shell being the most self-contained and easiest to understand for devs but depending on your existing Docker development flow, this might be the shortest path to multiple containers

**Comment by klieret** [comment]
As long as you have docker installed and running you can now run everything with just one line:
```bash
docker pull sweagent/swe-agent:latest && \
    docker run -it -v /var/run/docker.sock:/var/run/docker.sock \
    --rm sweagent/swe-agent-run:latest \
    python run.py --model_name gpt4 \
    --data_path https://github.com/klieret/swe-agent-test-repo/issues/1 \
    --config_file config/default_from_url.yaml  --skip_existing=False \
    --image_name=sweagent/swe-agent:latest
```
Admittedly, this is a very long command (I hope to shave off at least two lines tomorrow), but it should work on OSX and Linux and even on Windows (with one small modification). It’s gonna take a few minutes the first time you run it, but afterwards it should be as fast as with the normal installation.

Modification for windows: Use `//var/run/docker.sock:/var/run/docker.sock` (double leading `/` to avoid escaping, see [here](https://stackoverflow.com/questions/47225616/docker-inside-a-docker-on-windows)).

**Comment by klieret** [comment]
@cpendery Could you test on Windows? (I still plan to get to your PRs, but will probably only happen tomorrow now)

**Comment by klieret** [comment]
Ah, this still needs the injection of the keys

**Comment by cpendery** [comment]
We could always use WinGet and Homebrew to distribute the package. It would be easy to add docker & Python 3.9 as a dependency in the manifest/formula.

**Comment by cpendery** [comment]
I also have some install scripts I used with [kaldo](https://github.com/cpendery/kaldo) that could work here and they don't require sudo/admin installs on Unix or Windows. 

Docker has an [install script](https://github.com/docker/docker-install) or we could use the WinGet/brew packages in the script

**Comment by cpendery** [comment]
Actually what might be best is having user's install the package via `pipx` which will take care of the python component and then have an `init` subcommand, so the install instructions could look like

```
pipx install sweagent
sweagent init
```

I've made a sample init python script in #92 to demo the concept

**Comment by samuela** [comment]
You could also make keys.cfg optional and accept those values as env vars or CLI args as well

**Comment by klieret** [comment]
> You could also make keys.cfg optional and accept those values as env vars or CLI args as well
They can be specified as env vars. If you use docker, you have to pass them in with the `-e` flag

**Comment by klieret** [comment]
There's no experimental amd64 support for the docker image, see the instructions in #107

**Comment by RudolphSikorskiy** [comment]
> Feedback is appreciated!

Docker-compose solve that problem, does it?
It will only one command `docker-compose up` and all env vars with any adjustments will included

**Comment by RudolphSikorskiy** [comment]
```yaml
version: "3.9"

services:
  application:
    image: sweagent/swe-agent-run:latest
    restart: always
    volumes:
        - /var/run/docker.sock:/var/run/docker.sock
        - /keys.cfg:/app/keys.cfg
    environment:
      - GITHUB_TOKEN=$GITHUB_TOKEN
      - OPENAI_API_KEY=$OPENAI_API_KEY
      - DATA_PATH=$DATA_PATH
    command: ["python", "run.py", "--image_name=sweagent/swe-agent:latest", "--model_name", "gpt4", "--data_path", "https://github.com/pvlib/pvlib-python/issues/1603", "--config_file", "config/default_from_url.yaml", "--skip_existing=False"]
```

**Comment by klieret** [comment]
I honestly do not see how the docker compose script makes things easier. It just has lots of things hardcoded that shouldn't be (for example `keys.cfg` will have different paths), most of the command line arguments are user arguments, etc. So in the end, it's not much different than simply having a shell script with the command in it that any user can adapt.

**Comment by RudolphSikorskiy** [comment]
keys.cfg hardcoded in other cases, while application works end-to-end it does not matter. 
If you want service like approach, it should provide rest api with POST method and use it with parameters in payload. Responses better store in database or\and S3. And provide GET methods for use output data  to visualise it on frontend or smth else

**Comment by samuela** [comment]
> If you want service like approach, it should provide rest api with POST method and use it with parameters in payload. Responses better store in database or\and S3. And provide GET methods for use output data to visualise it on frontend or smth else

This is overly complex IMHO -- at least for the time being. The fundamental model here is a function `GitHub Issue -> PR`. Getting that function to work as smoothly and reliably as possible is a necessary first step towards anything else that might come later (API, caching, web interface, etc). And a CLI is just about the most ergonomic UX to represent a function `GitHub Issue -> PR`.

**Comment by klieret** [comment]
I'm going to close this now, as we've implemented some of this and the setup of codespaces #281 makes trying out the software even easier for people less familiar with the prerequisites. Let's open specific issues for new ideas.
